### PR TITLE
Feature/cloud

### PR DESCRIPTION
**Pull request by beatlevic**
Added cloud commands:

Screenshot 2020-12-07 at 22 28 47

Notes:
- Added the use of `execSync`, instead of the `util/shell` exec, which didn't handle double quotes
- Added `execSyncInherit` to allow live console output updates (pass through)
- Added the use of `child_process spawn` also with inherit, to allow login into remote containers with bash taking over

Still todo: 
- Update and warn secrets
- Allow deployment of feature-branches
- Update original documentation of deployment scripts

### CODE REVIEW

File: install [hunk 1]
```diff
@@ -41,6 +41,18 @@ _() {
       echo 'Error: Either yarn or npm must be installed (https://www.npmjs.com/)' >&2
       exit 1
     fi
+    if ! [ -x "$(command -v terraform)" ]; then
```
Comment (andrewplummer) [inline install:44]: Are we going to require all these as hard bedrock dependencies? I kind of think there will be pushback to this. Nearly everyone has git and node/npm installed but most will not have any of these, and I kind of doubt we want to make them install everything to try out bedrock quickly. Also what if they want to deploy to another platform, etc etc

File: src/cloud/index.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+  let labelName = service;
+  if (subservice) labelName += `-${subservice}`;
+  const podLabel = label || 'app';
+
+  const query = `resource.type="k8s_container"\nresource.labels.project_id="${project}"\nresource.labels.location="${computeZone}"\nresource.labels.cluster_name="${kubernetes.clusterName}"\nresource.labels.namespace_name="default"\nlabels.k8s-pod/${podLabel}="${labelName}"`;
+
+  const params = new URLSearchParams({ query });
+
+  const url = `https://console.cloud.google.com/logs/query;${params.toString()}?project=${project}`;
```
Comment (andrewplummer) [inline src/cloud/index.js:182]: this is a very odd url... is this how it's actually structured?

File: src/cloud/rollout.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  if (stat) tag += '-dirty';
+  return tag;
+}
+
+async function getMetaData() {
+  const author = await exec('git config user.name');
+  const date = new Date().toUTCString();
+  const branch = await exec('git branch --show-current');
+  const git = await getTag();
+  let fields = `\\"author\\":\\"${author}\\"`;
```
Comment (andrewplummer) [inline src/cloud/rollout.js:26]: could this not just be `JSON.stringify`?

File: src/cloud/rollout.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    environment,
+    'services',
+    `${deployment}.yml`
+  );
+
+  // Check for config file as it might not exist if the
+  // deployment was dynamically created for a feature branch.
+  if (fs.existsSync(deploymentFile)) {
+    const applyCommand = `kubectl apply -f ${deploymentFile} --record`;
+    // console.info(applyCommand);
```
Comment (andrewplummer) [inline src/cloud/rollout.js:50]: remove?

File: src/util/sleep.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export async function sleep(ms) {
```
Comment (andrewplummer) [inline src/util/sleep.js:1]: is this being used?

File: src/util/shell.js [hunk 6]
```diff
@@ -1,19 +1,29 @@
...
     return output;
-  } catch(err) {
+  } catch (err) {
     const { stderr, stdout } = err;
     const message = (stderr || stdout).split('\n').join(' ');
     throw new Error(message);
   }
 }
+
+export async function execSyncInherit(command) {
```
Comment (andrewplummer) [inline src/util/shell.js:23]: I'm not familiar with `stdio: inherit` ... is there any way we could fold this into `exec`? If not maybe document when/how to use this function?

File: src/cloud/index.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    process.exit(0);
+  }
+
+  const podName = filteredPods[0].metadata.name;
+  console.info(kleur.green(`=> Starting bash for pod: "${podName}"`));
+
+  const { spawn } = require('child_process');
+
+  const child = spawn('kubectl', ['exec', '-it', podName, '--', 'bash'], {
+    stdio: 'inherit',
```
Comment (andrewplummer) [inline src/cloud/index.js:158]: I'm assuming that opening a shell into a pod is a one off thing that goes beyond what `exec` can handle, so it needs to be done custom like this?

**Comment by andrewplummer** [comment]
@beatlevic is it not possible to fix `exec` to handle the double quotes?

File: install (hunk 1 above)
Comment (beatlevic) [reply install:44]: Good point. I'll move this check to when you actually try to run a `cloud` command.

File: src/cloud/index.js (hunk 2 above)
Comment (beatlevic) [reply src/cloud/index.js:182]: Don't get me started, ask Google :) This is the URL I see when I'm browsing the Google Logs Explorer.

File: install (hunk 1 above)
Comment (andrewplummer) [reply install:44]: also maybe a note when they run `bedrock cloud` command? (if it's not there already) ... I think you can update the `description` field in `commands.json`.

File: src/cloud/rollout.js (hunk 3 above)
Comment (beatlevic) [reply src/cloud/rollout.js:26]: That's what I started out with, but unfortunately no. This relates to the issue with `exec` and double quotes.

File: src/cloud/rollout.js (hunk 4 above)
Comment (beatlevic) [reply src/cloud/rollout.js:50]: I'll remove the comment. However, do we want to output the command that will run for the user to see?

File: src/util/sleep.js (hunk 5 above)
Comment (beatlevic) [reply src/util/sleep.js:1]: Not anymore. I sometimes use it when testing.

File: src/cloud/index.js (hunk 7 above)
Comment (beatlevic) [reply src/cloud/index.js:158]: Yes, you basically hand off TTY control to the spawend process.

File: src/util/shell.js (hunk 6 above)
Comment (beatlevic) [reply src/util/shell.js:23]: For example, when you are pushing your docker image to gcr.io, you see progression animations, that you wouldn't see if you use the `exec` that only captures the end result (final output string). stdio `inherit` does a pass through of the output to the current (node.js) process, so you can see the live progression/status output.

I'll add documentation.

File: install (hunk 1 above)
Comment (beatlevic) [reply install:44]: I mean, `cloud provision` command.

File: src/cloud/rollout.js (hunk 3 above)
Comment (andrewplummer) [reply src/cloud/rollout.js:26]: ahh of course I get it now..... but maybe just `JSON.stringify(...).replace(/"/g, "\\")`? maybe exec could even do it?

File: src/cloud/rollout.js (hunk 4 above)
Comment (andrewplummer) [reply src/cloud/rollout.js:50]: not sure if it works for you but `runAsTask` is meant to be the way to run a one-off task with a loader and a message to the user (including error message on fail)

File: src/util/sleep.js (hunk 5 above)
Comment (andrewplummer) [reply src/util/sleep.js:1]: ahh ok either way

File: src/util/shell.js (hunk 6 above)
Comment (andrewplummer) [reply src/util/shell.js:23]: ahh ok this totally makes sense then... I guess `cloud` will be fundamentally different from other commands where the user is going to actually care about the details of what's going on whereas other commands we can/should hide the bedrock-specific details...

File: src/cloud/index.js (hunk 7 above)
Comment (andrewplummer) [reply src/cloud/index.js:158]: cool makes sense 👍

**Comment by dominiek** [comment]
Error when getting `info`:

```bash
dodo@mini-gibson ~/c/bedrock-core> bedrock cloud info staging api
Using Google Cloud environment staging (project=bedrock-foundation, compute/zone=us-east1-c, cluster=cluster-1)
(node:17741) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token m in JSON at position 0
    at JSON.parse (<anonymous>)
    at checkDeployment (/Users/dodo/.bedrock/src/cloud/rollout.js:93:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async info (/Users/dodo/.bedrock/src/cloud/index.js:110:26)
    at async Command.<anonymous> (/Users/dodo/.bedrock/bin/bedrock:74:9)
(node:17741) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:17741) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by dominiek** [comment]
Error when running `shell` command:

```bash
dodo@mini-gibson ~/c/bedrock-core> bedrock cloud shell staging api
Using Google Cloud environment staging (project=bedrock-foundation, compute/zone=us-east1-c, cluster=cluster-1)
(node:18030) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token m in JSON at position 1
    at JSON.parse (<anonymous>)
    at shell (/Users/dodo/.bedrock/src/cloud/index.js:138:21)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Command.<anonymous> (/Users/dodo/.bedrock/bin/bedrock:74:9)
(node:18030) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:18030) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by dominiek** [comment]
`bedrock cloud staging logs mongo` does not seem to work.

**Comment by dominiek** [comment]
- Detect the version of kube-ctl
- Tell the user to upgrade (`brew upgrade kubectl`) if version is old

**Comment by dominiek** [comment]
This is a known issue, but adding it here:

```
$ cd e-flux
$ gcloud config configurations activate e-flux
$ ./deployment/scripts/authorize staging
$ bedrock cloud status staging
  Invalid Google Cloud config (use authorize script): project = eflux-staging
  Invalid Google Cloud config (use authorize script): container/cluster = standard-cluster-1
$ gcloud config configurations activate bedrock
$ bedrock cloud status staging
  SHOWS E-FLUX STATUS
```

**Comment by dominiek** [comment]
When service or environment doesn't exist: give a nice error

**Review by dominiek — CHANGES_REQUESTED** [review]
- Checking for existence of service or environment
- Ensuring right Kubernetes version is installed
- Deployment configuration on bedrock-core PR
- A reference to the deployment playbook from the README in this repo

File: src/cloud/rollout.js (hunk 3 above)
Comment (beatlevic) [reply src/cloud/rollout.js]: Fixed

**Comment by beatlevic** [comment]
> * Detect the version of kube-ctl
> * Tell the user to upgrade (`brew upgrade kubectl`) if version is old

Is now implemented.

**Comment by beatlevic** [comment]
> When service or environment doesn't exist: give a nice error

Is now Implemented.

**Review by dominiek — APPROVED** [review]

**Pull request merged**